tutorials/haskell/hspec-golden-test-tutorial/tutorial.md

```diff
@@ -0,0 +1,506 @@
+      Golden and Actual output hasn't changed
+Hello.Hello
+  sayHi
+    shows a Hello Golden Testers string
+Html.HtmlGolden
+  renderHtml
+    Renders an Html5 file 
+      Golden and Actual output hasn't changed
+
+Finished in 0.0034 seconds
+3 examples, 0 failures
+```
+
+And we can see that everything works fine.
+
+#### JSON Golden Test
+
+Until now we have managed components that returns strings only, but what if our return type is different like a ByteString or Text? For example, lets analize the `encodeCountries` function in the `Json.hs` module.
+
+```Haskell
+encodeCountries :: [Country] -> ByteString
+encodeCountries = encode
+```
+
+This function returns a ByteString, that means that if we try to assert the output with our `defaultGolden` function,
+
+``` Haskell
+  describe "encodeCountries" $ do
+   it "encodes a group of Countries into a JSON String " $
+    defaultGolden "json" (encodeCountries countries)
+```
+
+test won't even compile.
+
+```shell
+$stack test
+• Couldn't match type ‘B.ByteString’ with ‘[Char]’
+      Expected type: String
+        Actual type: B.ByteString
+```
+
+Conviniently `hspec-golden` exports a `Golden` data type. With this tool we can configure our Golden test 
+specifying the functions `hspec-golden` should use for our test to work.
+
+```
+data Golden str =
+  Golden {
+    output       :: str,                      --  Output
+    encodePretty :: str -> String,            --  Makes the comparison pretty when the test fails
+    writeToFile  :: FilePath -> str -> IO (), --  How to write into the golden file the file
+    readFromFile :: FilePath -> IO str,       --  How to read the file,
+    testName     :: String,                   --  Test name 
+    directory    :: FilePath                  --  Directory where you write your tests
+  }
+```
+For example we can define a different directory for storing our golden files or we can define how our test will be read and how the output should be written. Also the `encodePretty` characteristic determines how the prompt should print a readable output when tests fails. With this data-type we are now able to create a new function to assert our output. Let's start by importing some modules.
+
+``` Haskell
+module Json.JsonGoldenSpec where
+
+import           Test.Hspec
+import           Test.Hspec.Golden
+import           JSON.Json
+import qualified Data.ByteString.Lazy as B    
+```
+
+Next, based on the `Golden` data type documentation , lets create our assert function.
+
+```Haskell
+goldenBytestring :: String -> B.ByteString -> Golden B.ByteString
+goldenBytestring name actualOutput =
+    Golden {
+        output = actualOutput,
+        encodePretty = show,
+        writeToFile = B.writeFile,
+        readFromFile = B.readFile,
+        testName = name,
+        directory = ".otherGolden"
+          
+    }
+```
+
+Let's analyze a little our `goldenBytestring` function: 
+
+ - It return type is a `Golden ByteString`.
```

Comment:
@Jagl257 _Its type is `Golden ByteString`_ (?)

Reply:
@CristhianMotoche sounds much better, thanks.